File: linux-client/osm_ride_linux/util/units.py
"""Formatting helpers - the app displays imperial units throughout (distance, weight), except
waist/neck/height, which are shown in centimeters per user preference even though they're stored
in the same unit anyway, so there's no conversion to do for those.

Mirrors app/src/main/java/com/ewaldmire/osmride/util/Units.kt.
"""

from __future__ import annotations

_METERS_PER_MILE = 1609.344
_METERS_PER_FOOT = 0.3048
_MPS_TO_MPH = 2.2369362921
_LB_PER_KG = 2.2046226218


def meters_to_miles(meters: float) -> float:
    return meters / _METERS_PER_MILE


def meters_to_feet(meters: float) -> float:
    return meters / _METERS_PER_FOOT


def mps_to_mph(meters_per_second: float) -> float:
    return meters_per_second * _MPS_TO_MPH


def kg_to_lbs(kg: float) -> float:
    return kg * _LB_PER_KG


def lbs_to_kg(lbs: float) -> float:
    return lbs / _LB_PER_KG


def format_miles(meters: float) -> str:
    return f"{meters_to_miles(meters):.2f} mi"


def format_feet(meters: float) -> str:
    return f"{meters_to_feet(meters):.0f} ft"


def format_mph(meters_per_second: float) -> str:
    return f"{mps_to_mph(meters_per_second):.1f} mph"


def format_watts(watts: float | None) -> str:
    return "--" if watts is None else f"{watts:.0f} W"


def format_cadence(rpm: float | None) -> str:
    return "--" if rpm is None else f"{rpm:.0f} rpm"


def format_heart_rate(bpm: float | int | None) -> str:
    return "--" if bpm is None else f"{bpm:.0f} bpm"


def format_grade(percent: float | None) -> str:
    return "--" if percent is None else f"{percent:+.1f}%"


def format_kilocalories(kcal: float | None) -> str:
    return "--" if kcal is None else f"{kcal:.0f} Cal"


def format_weight_lbs(kg: float) -> str:
    return f"{kg_to_lbs(kg):.1f} lb"


def format_waist_cm(cm: float) -> str:
    return f"{cm:.1f} cm"


def format_body_fat_percent(percent: float | None) -> str:
    return "--" if percent is None else f"{percent:.1f}%"


def format_duration(total_seconds: float) -> str:
    total_seconds = int(total_seconds)
    h = total_seconds // 3600
    m = (total_seconds % 3600) // 60
    s = total_seconds % 60
    if h > 0:
        return f"{h}:{m:02d}:{s:02d}"
    return f"{m}:{s:02d}"

File: app/src/main/java/com/ewaldmire/osmride/util/Units.kt
package com.ewaldmire.osmride.util

import java.util.Locale

/** Formatting helpers — the app displays imperial units throughout, per how it's used. */
object Units {
    private const val METERS_PER_MILE = 1609.344
    private const val METERS_PER_FOOT = 0.3048
    private const val MPS_TO_MPH = 2.2369362921
    private const val LB_PER_KG = 2.2046226218
    private const val CM_PER_INCH = 2.54

    fun metersToMiles(meters: Double): Double = meters / METERS_PER_MILE
    fun metersToFeet(meters: Double): Double = meters / METERS_PER_FOOT
    fun mpsToMph(metersPerSecond: Double): Double = metersPerSecond * MPS_TO_MPH
    fun kgToLbs(kg: Double): Double = kg * LB_PER_KG
    fun lbsToKg(lbs: Double): Double = lbs / LB_PER_KG
    fun cmToInches(cm: Double): Double = cm / CM_PER_INCH
    fun inchesToCm(inches: Double): Double = inches * CM_PER_INCH

    fun formatMiles(meters: Double): String = String.format(Locale.US, "%.2f mi", metersToMiles(meters))
    fun formatFeet(meters: Double): String = String.format(Locale.US, "%.0f ft", metersToFeet(meters))
    fun formatMph(metersPerSecond: Double): String = String.format(Locale.US, "%.1f mph", mpsToMph(metersPerSecond))
    fun formatWatts(watts: Double?): String = if (watts == null) "--" else String.format(Locale.US, "%.0f W", watts)
    fun formatCadence(rpm: Double?): String = if (rpm == null) "--" else String.format(Locale.US, "%.0f rpm", rpm)
    fun formatHeartRate(bpm: Double?): String = if (bpm == null) "--" else String.format(Locale.US, "%.0f bpm", bpm)
    fun formatHeartRate(bpm: Int?): String = if (bpm == null) "--" else "$bpm bpm"
    fun formatGrade(percent: Double?): String =
        if (percent == null) "--" else String.format(Locale.US, "%+.1f%%", percent)
    fun formatKilocalories(kcal: Double?): String =
        if (kcal == null) "--" else String.format(Locale.US, "%.0f Cal", kcal)
    fun formatWeightLbs(kg: Double): String = String.format(Locale.US, "%.1f lb", kgToLbs(kg))
    fun formatWaistInches(cm: Double): String = String.format(Locale.US, "%.1f in", cmToInches(cm))
    fun formatBodyFatPercent(percent: Double?): String =
        if (percent == null) "--" else String.format(Locale.US, "%.1f%%", percent)

    fun formatDuration(totalSeconds: Long): String {
        val h = totalSeconds / 3600
        val m = (totalSeconds % 3600) / 60
        val s = totalSeconds % 60
        return if (h > 0) {
            String.format(Locale.US, "%d:%02d:%02d", h, m, s)
        } else {
            String.format(Locale.US, "%d:%02d", m, s)
        }
    }
}
